src/rgw/driver/rados/rgw_sal_rados.cc

```diff
@@ -4624,6 +4632,283 @@ std::unique_ptr<LCSerializer> RadosLifecycle::get_serializer(const std::string&
+
+  try {
+    fifos[index]->push(dpp, std::move(bl), ceph::async::use_blocked);
+  } catch (const sys::system_error& e) {
+    ldpp_dout(dpp, -1) << __PRETTY_FUNCTION__
+		 << ": unable to push to FIFO: " << obj_names[index]
+		 << ": " << e.what() << dendl;
+    throw;
+  }
+  return 0;
+}
+
+struct rgw_restore_fifo_entry {
+  std::string id;
+  ceph::real_time mtime;
+  RGWRestoreEntry entry;
+  rgw_restore_fifo_entry() {}
+
+  void encode(ceph::buffer::list& bl) const {
+    ENCODE_START(1, 1, bl);
+    encode(id, bl);
+    encode(mtime, bl);
+    encode(entry, bl);
+    ENCODE_FINISH(bl);
+  }
+
+  void decode(ceph::buffer::list::const_iterator& bl) {
+     DECODE_START(1, bl);
+     decode(id, bl);
+     decode(mtime, bl);
+     decode(entry, bl);
+     DECODE_FINISH(bl);
+  }
+
+  void dump(ceph::Formatter* f) const;
+  void decode_json(JSONObj* obj);
+};
+WRITE_CLASS_ENCODER(rgw_restore_fifo_entry)
+
+int RadosRestore::list(const DoutPrefixProvider *dpp, optional_yield y,
+	       	   int index, const std::string& marker, std::string* out_marker,
+		   uint32_t max_entries, std::vector<RGWRestoreEntry>& entries,
+		   bool* truncated)
+{
+  std::vector<fifo::entry> restore_entries{max_entries};
+  std::string omark = {};
+  bool more = false;
+
+  ldpp_dout(dpp, 20) << __PRETTY_FUNCTION__
+		 << "Listing entries from FIFO:" << obj_names[index] << dendl;
+
+  try {
+    auto [lentries, lmark] = fifos[index]->list(dpp, marker,
+			  	 restore_entries, ceph::async::use_blocked);
+    entries.clear();
+
+    for (const auto& entry : lentries) {
+      rgw_restore_fifo_entry r_entry;
+      r_entry.id = entry.marker;
+      r_entry.mtime = entry.mtime;
+
+      auto liter = entry.data.cbegin();
+      try {
+	decode(r_entry.entry, liter);
+      } catch (const buffer::error& err) {
+	ldpp_dout(dpp, -1) << __PRETTY_FUNCTION__
+		   << ": failed to decode restore entry: "
+		   << err.what() << dendl;
+	return -EIO;
+      }
+      RGWRestoreEntry& e = r_entry.entry;
+      entries.push_back(std::move(e));
+      omark = entry.marker;
+    }
+    if (!lmark.empty()) {
+      more = true;
+    }
+  } catch (const sys::system_error& e) {
+    if (e.code() == sys::errc::no_such_file_or_directory) {
+    } else {
+      ldpp_dout(dpp, -1) << __PRETTY_FUNCTION__
+		 << ": unable to list FIFO: " << obj_names[index]
+		 << ": " << e.what() << dendl;
+      throw;
+    }
+  }
+
+  if (truncated) {
+    *truncated = more;
+  }				 
+
+  if (out_marker) {
+    *out_marker = omark;
+  }
+
+  ldpp_dout(dpp, 20) << __PRETTY_FUNCTION__
+		 << "Listing from FIFO:" << obj_names[index] << ", returned:"
+		 << restore_entries.size() << " entries, truncated:"
+		 << (truncated ? *truncated : false) 
+		 << ", out_marker:" << (out_marker ? *out_marker : "") << dendl;
+
+  return 0;
+}
+
+int RadosRestore::trim_entries(const DoutPrefixProvider *dpp, optional_yield y,
+			int index, const std::string_view& marker)
+{
+  assert(index < num_objs);
```

Comment:
If this is something that can be called from an HTTP interface or radosgw it should return an error rather than asserting.

If it's called by things otherwise in the same process that should be validating their inputs, this is fine.